File: api/cruzar-inventario.js
import ExcelJS from "exceljs";
import formidable from "formidable";

export const config = {
  api: { bodyParser: false },
};

/* =========================
   EXTRAER VALOR REAL CELDA
========================= */
function obtenerValorCelda(cell) {
  if (!cell || cell.value == null) return "";

  const v = cell.value;

  if (typeof v === "string" || typeof v === "number") {
    return String(v);
  }

  if (v.text) return String(v.text);

  if (v.richText) {
    return v.richText.map(r => r.text).join("");
  }

  if (v.result) return String(v.result);

  return String(v);
}

/* =========================
   NORMALIZADOR
========================= */
function normalizar(valor) {
  return String(valor)
    .toLowerCase()
    .normalize("NFD")
    .replace(/[\u0300-\u036f]/g, "")
    .replace(/\s/g, "")
    .replace(/[^a-z0-9]/gi, "")
    .trim();
}

/* =========================
   HANDLER
========================= */
export default async function handler(req, res) {

  if (req.method !== "POST") {
    return res.status(405).json({ error: "Método no permitido" });
  }

  const form = formidable({ multiples: true });

  form.parse(req, async (err, fields, files) => {

    if (err) return res.status(500).json({ error: err.message });

    const inventarioFile = files.inventario?.[0];
    const escaneoFile = files.escaneo?.[0];

    if (!inventarioFile || !escaneoFile) {
      return res.status(400).json({ error: "Faltan archivos Excel" });
    }

    try {

      const wbInventario = new ExcelJS.Workbook();
      const wbEscaneo = new ExcelJS.Workbook();

      await wbInventario.xlsx.readFile(inventarioFile.filepath);
      await wbEscaneo.xlsx.readFile(escaneoFile.filepath);

      const wsInventario = wbInventario.worksheets[0];
      const wsEscaneo = wbEscaneo.worksheets[0];

      /* =========================
         EXTRAER CÓDIGOS ESCANEADOS
      ========================= */
      const codigosEscaneados = new Set();
      const codigosOriginales = new Map();
      let totalEscaneados = 0;

      wsEscaneo.eachRow((row) => {
        row.eachCell((cell) => {
          const valor = obtenerValorCelda(cell);
          const limpio = normalizar(valor);

          if (limpio) {
            codigosEscaneados.add(limpio);
            codigosOriginales.set(limpio, valor);
            totalEscaneados++;
          }
        });
      });

      if (codigosEscaneados.size === 0) {
        return res.status(400).json({ error: "El archivo de escaneo no tiene códigos válidos" });
      }

      /* =========================
         CRUCE TOTAL SIN COLUMNAS
      ========================= */
      let coincidencias = 0;
      const encontrados = new Set();

      wsInventario.eachRow((row) => {
        row.eachCell((cell) => {

          const valor = obtenerValorCelda(cell);
          const limpio = normalizar(valor);

          if (!limpio) return;

          if (codigosEscaneados.has(limpio)) {

            cell.fill = {
              type: "pattern",
              pattern: "solid",
              fgColor: { argb: "FF00FF00" }
            };

            coincidencias++;
            encontrados.add(limpio);
          }
        });
      });

      /* =========================
         CÓDIGOS NO ENCONTRADOS
      ========================= */
      const noEncontrados = [...codigosEscaneados].filter(c => !encontrados.has(c));

      if (noEncontrados.length > 0) {

        const inicio = wsInventario.rowCount + 2;

        wsInventario.getCell(`A${inicio}`).value =
          "CODIGOS ESCANEADOS QUE NO EXISTEN EN INVENTARIO";

        noEncontrados.forEach((codigo, index) => {
          wsInventario.getCell(`A${inicio + index + 1}`).value =
            codigosOriginales.get(codigo);
        });
      }

      /* =========================
         GENERAR RESULTADO
      ========================= */
      const buffer = await wbInventario.xlsx.writeBuffer();

      res.setHeader(
        "Content-Type",
        "application/vnd.openxmlformats-officedocument.spreadsheetml.sheet"
      );

      res.setHeader(
        "Content-Disposition",
        "attachment; filename=inventario_cruzado.xlsx"
      );

      res.send(buffer);

      console.log(
        `De ${totalEscaneados} códigos escaneados se hallaron ${coincidencias} coincidencias`
      );

    } catch (error) {

      return res.status(500).json({
        error: "Error procesando archivos",
        detalle: error.message
      });

    }

  });

}

File: api/cruzar_inventario.py

import io
from flask import Flask, request, send_file, jsonify
import pandas as pd
from openpyxl import load_workbook
from openpyxl.styles import PatternFill

app = Flask(__name__)

@app.route("/api/cruzar_inventario", methods=["POST"])
def cruzar_inventario():
    if "inventario" not in request.files or "escaneo" not in request.files:
        return jsonify({"error": "Faltan archivos"}), 400

    inventario_file = request.files["inventario"]
    escaneo_file = request.files["escaneo"]

    try:
        # Leer Excel con pandas
        df_inventario = pd.read_excel(inventario_file)
        df_escaneo = pd.read_excel(escaneo_file)

        # Normalizar columna 'codigo'
        df_inventario['codigo'] = df_inventario['codigo'].astype(str).str.strip()
        df_escaneo['codigo'] = df_escaneo['codigo'].astype(str).str.strip()

        # Crear set de códigos escaneados
        codigos_escaneo = set(df_escaneo['codigo'])

        # Guardar inventario temporalmente en memoria
        output_buffer = io.BytesIO()
        df_inventario.to_excel(output_buffer, index=False)
        output_buffer.seek(0)

        # Abrir con openpyxl para marcar en verde
        wb = load_workbook(output_buffer)
        ws = wb.active

        # Color verde
        verde = PatternFill(start_color="00FF00", end_color="00FF00", fill_type="solid")

        # Buscar columna 'codigo'
        col_codigo = None
        for col in range(1, ws.max_column + 1):
            if ws.cell(row=1, column=col).value == "codigo":
                col_codigo = col
                break
        if col_codigo is None:
            return jsonify({"error": "No se encontró la columna 'codigo'"}), 400

        # Marcar coincidencias exactas
        coincidencias = 0
        for row in range(2, ws.max_row + 1):
            codigo = str(ws.cell(row=row, column=col_codigo).value).strip()
            if codigo in codigos_escaneo:
                ws.cell(row=row, column=col_codigo).fill = verde
                coincidencias += 1

        # Guardar Excel final en memoria
        final_buffer = io.BytesIO()
        wb.save(final_buffer)
        final_buffer.seek(0)

        # Devolver Excel como attachment + header con coincidencias
        return send_file(
            final_buffer,
            as_attachment=True,
            download_name="inventario_cruzado.xlsx",
            mimetype="application/vnd.openxmlformats-officedocument.spreadsheetml.sheet",
            headers={"X-Coincidencias": str(coincidencias)}
        )

    except Exception as e:
        return jsonify({"error": str(e)}), 500

if __name__ == "__main__":
    app.run(debug=True)
